tests/integration/test_tracing.py

```diff
@@ -535,6 +596,228 @@ def tracking_function():
+        thread.join()
+
+    # THEN
+    allocations = list(FileReader(output).get_allocation_records())
+
+    vallocs = [
+        event
+        for event in allocations
+        if event.size == 1234 and event.allocator == AllocatorType.VALLOC
+    ]
+    assert len(vallocs) == 2
+
+    first, second = vallocs
+    alloc1_funcs = [frame[0] for frame in first.stack_trace()]
+    alloc2_funcs = [frame[0] for frame in second.stack_trace()]
+
+    # Cython frames called before tracking started aren't in the Python stack.
+    assert alloc1_funcs[:3] == ["valloc", "func1", "thread_body"]
+    assert alloc2_funcs[:2] == ["valloc", "thread_body"]
+
+
+def test_cython_frame_in_pre_existing_thread_stack_when_restarting_tracking(tmp_path):
+    """Test restarting tracking when another thread's stack has Cython frames.
+
+    The intended execution flow is:
+    Main Thread          Background Thread
+    -----------          -----------------
+    Install tracker
+    Start thread
+                         Call thread_body
+                         Call function_caller (Cython frame)
+                         Call func1
+    Uninstall tracker
+    Install new tracker
+                         Perform an allocation
+                         Return from func1
+                         Return from function_caller
+                         Perform an allocation
+                         Return from thread_body
+    Join thread
+    Uninstall tracker
+    """
+    # GIVEN
+    ready_to_replace_tracker = threading.Event()
+    tracker_replaced = threading.Event()
+
+    allocator = MemoryAllocator()
+    output1 = tmp_path / "test.bin.1"
+    output2 = tmp_path / "test.bin.2"
+
+    def thread_body():
+        function_caller(func1)
+        allocator.valloc(1234)
+        allocator.free()
+
+    def func1():
+        ready_to_replace_tracker.set()
+        tracker_replaced.wait()
+        allocator.valloc(1234)
+        allocator.free()
+
+    # WHEN
+    with Tracker(output1):
+        thread = threading.Thread(target=thread_body)
+        thread.start()
+        ready_to_replace_tracker.wait()
+
+    with Tracker(output2):
+        tracker_replaced.set()
+        thread.join()
+
+    # THEN
+    allocations = list(FileReader(output2).get_allocation_records())
+
+    vallocs = [
+        event
+        for event in allocations
+        if event.size == 1234 and event.allocator == AllocatorType.VALLOC
+    ]
+    assert len(vallocs) == 2
+
+    first, second = vallocs
+    alloc1_funcs = [frame[0] for frame in first.stack_trace()]
+    alloc2_funcs = [frame[0] for frame in second.stack_trace()]
+
+    # Cython frames called before tracking started aren't in the Python stack.
+    assert alloc1_funcs[:3] == ["valloc", "func1", "thread_body"]
+
+    # But we do get pop events for them if tracking was enabled when they were
+    # pushed. Ensure we handle these unexpected pops.
+    assert alloc2_funcs[:2] == ["valloc", "thread_body"]
+
+
```

Comment:
Would be great if we could add a test that's doing some allocations in native code before the tracker gets installed

Reply:
I don't follow. If it's doing allocations before the tracker gets installed, we won't see those allocations, because our malloc hooks aren't installed yet. What would the test be looking for?

Reply:
Pablo and I chatted offline, and he explained his goal in a bit more detail. He wants a test that sets up a Python call stack and then drops the GIL in C code. While the GIL is dropped, another thread installs a tracker. After the tracker is installed, the first thread performs an allocation, without reacquiring the GIL. We expect the Python stack associated with that allocation to be correct, but no test currently covers that case. I'll add one.